From 8b925d95e062df70a7046570ac70e169448d61ad Mon Sep 17 00:00:00 2001
From: Geetha sowjanya <gakula@marvell.com>
Date: Tue, 9 Feb 2021 17:44:04 +0530
Subject: [PATCH 702/767] octeontx2-af: Use pci revision id for platform check.

commit ad39c83debb0dd58e0fde566d76293dd172c1586 from
git@git.assembla.com:cavium/WindRiver.linux.git

Due to the hardware bug on octeontx2 platforms, pci
subsystem ids of RVU devices populates incorrectly.
Hence, use pci revision ids for the platform check.

Change-Id: I6e275ca4b814f29684734ffc092e8f212010c62b
Signed-off-by: Geetha sowjanya <gakula@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/45686
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 .../net/ethernet/marvell/octeontx2/af/rvu.h   | 23 +++++++++++++------
 1 file changed, 16 insertions(+), 7 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu.h b/drivers/net/ethernet/marvell/octeontx2/af/rvu.h
index 6ffa090bfd3c..0d980c5f425f 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu.h
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu.h
@@ -24,9 +24,6 @@
 /* Subsystem Device ID */
 #define PCI_SUBSYS_DEVID_98XX                  0xB100
 #define PCI_SUBSYS_DEVID_96XX                  0xB200
-#define PCI_SUBSYS_DEVID_95XX                  0xB300
-#define PCI_SUBSYS_DEVID_LOKI                  0xB400
-#define PCI_SUBSYS_DEVID_95XXMM                0xB500
 
 /* PCI BAR nos */
 #define	PCI_AF_REG_BAR_NUM			0
@@ -544,13 +541,25 @@ static inline bool is_rvu_95xx_A0(struct rvu *rvu)
 	return (pdev->revision == 0x10) || (pdev->revision == 0x11);
 }
 
+/* REVID for PCIe devices.
+ * Bits 0..1: minor pass, bit 3..2: major pass
+ * bits 7..4: midr id
+ */
+#define PCI_REVISION_ID_96XX		0x00
+#define PCI_REVISION_ID_95XX		0x10
+#define PCI_REVISION_ID_LOKI		0x20
+#define PCI_REVISION_ID_98XX		0x30
+#define PCI_REVISION_ID_95XXMM		0x40
+
 static inline bool is_rvu_otx2(struct rvu *rvu)
 {
-	unsigned short id = rvu->pdev->subsystem_device;
+	struct pci_dev *pdev = rvu->pdev;
+
+	u8 midr = pdev->revision & 0xF0;
 
-	return (id == PCI_SUBSYS_DEVID_96XX || id == PCI_SUBSYS_DEVID_98XX ||
-		id == PCI_SUBSYS_DEVID_95XX || id == PCI_SUBSYS_DEVID_LOKI ||
-		id == PCI_SUBSYS_DEVID_95XXMM);
+	return (midr == PCI_REVISION_ID_96XX || midr == PCI_REVISION_ID_95XX ||
+		midr == PCI_REVISION_ID_LOKI || midr == PCI_REVISION_ID_98XX ||
+		midr == PCI_REVISION_ID_95XXMM);
 }
 
 static inline bool is_cgx_mapped_to_nix(unsigned short id, u8 cgx_id)
-- 
2.31.1

